tutorials/accelerated-python/notebooks/fundamentals/03__numpy_to_cupy__ndarray_basics.ipynb

```diff
@@ -1,385 +1,509 @@
+      "source": [
+        "with cp.cuda.Device(0):\n",
+        "   x_on_gpu0 = cp.random.random((100000, 1000))\n",
+        "\n",
+        "print(f\"Array is on device: {x_on_gpu0.device}\")"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "id": "32f7226a",
+      "metadata": {},
+      "source": [
+        "**Note:** CuPy functions generally expect all input arrays to be on the **same** device. Passing an array stored on a non-current device may work depending on the hardware configuration but is generally discouraged as it may not be performant.\n"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "id": "2e0a4a03",
+      "metadata": {},
+      "source": [
+        "---"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "id": "d2e341ff-0c1e-40e8-8c33-9e3039de8013",
+      "metadata": {
+        "id": "d2e341ff-0c1e-40e8-8c33-9e3039de8013"
+      },
+      "source": [
+        "## Exercise - NumPy to CuPy\n",
+        "\n",
+        "### Part 1\n",
+        "Let's put the \"Drop-in Replacement\" philosophy to the test with the same data pipeline as the previous notebook. Specficially, the single block of code below performs the following steps:\n",
+        "1) Generate a massive dataset (50 million elements).\n",
+        "2) Process it using a heavy operation (Sorting).\n",
+        "3) Manipulate the shape and normalize the data (Broadcasting).\n",
+        "4) Verify the integrity of the result.\n",
+        "\n",
+        "**TODO:**\n",
+        "1. Run the cell below with xp = np (CPU Mode). Note the \"Sort Time\".\n",
+        "2. Change the setup line to xp = cp (GPU Mode). Run it again.\n",
+        "3. Observe how the exact same logic runs significantly faster on the GPU with CuPy while retaining the implementation properties of NumPy."
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "id": "cc4596d8-d9ff-4c66-8822-246c0fc830c7",
+      "metadata": {
+        "id": "cc4596d8-d9ff-4c66-8822-246c0fc830c7"
+      },
+      "outputs": [],
+      "source": [
+        "import numpy as np\n",
+        "import cupy as cp\n",
+        "import time\n",
+        "\n",
+        "# --- 1. SETUP: CHOOSE YOUR DEVICE ---\n",
+        "xp = np  # Toggle this to 'cp' for GPU acceleration\n",
+        "\n",
+        "print(f\"Running on: {xp.__name__.upper()}\")\n",
+        "\n",
+        "# --- 2. DATA GENERATION ---\n",
+        "N = 50_000_000\n",
+        "print(f\"Generating {N:,} random elements ({N*8/1e9:.2f} GB)...\")\n",
+        "arr = xp.random.rand(N)\n",
+        "\n",
+        "# --- 3. HEAVY COMPUTATION (TIMED) ---\n",
+        "print(\"Sorting data...\")\n",
+        "t0 = time.perf_counter()\n",
+        "\n",
+        "xp.sort(arr)\n",
+        "\n",
+        "# Ensure GPU finishes before stopping timer\n",
+        "if xp == cp:\n",
+        "    cp.cuda.Stream.null.synchronize()\n",
+        "\n",
+        "t1 = time.perf_counter()\n",
```

Comment:
I'm a littel concerned about using %%timeit, %time, and perf_counter. They require us to add stream synchronizes, which means we have to explain them, and it's easy to forget them.

Why not use CuPy's benchmarking facility?